From 03f0121e214790bac11140ef6c7d534bf2eba5a1 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Wed, 10 Apr 2024 18:57:59 +0200
Subject: [PATCH 02/48] === AVPro Video seeking support ===

---
 MAINTAINERS | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index c50c5d235c9..2a76a529c7f 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -416,3 +416,5 @@ P:	Elizabeth Figura <zfigura@codeweavers.com>
 P:	Paul Gofman <gofmanp@gmail.com>
 P:	Erich E. Hoover <erich.e.hoover@wine-staging.com>
 W:	https://wine-staging.com/
+
+seek
-- 
2.51.0

