From 05993457227e1e9aa4f7018bddfb152d122efd43 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Wed, 31 Jan 2024 17:42:55 +0100
Subject: [PATCH 10/19] winegstreamer: Set base time on wg_parser bin while
 connecting.

---
 dlls/winegstreamer/wg_parser.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/winegstreamer/wg_parser.c b/dlls/winegstreamer/wg_parser.c
index 4f863d40bfc..fef8ae883b3 100644
--- a/dlls/winegstreamer/wg_parser.c
+++ b/dlls/winegstreamer/wg_parser.c
@@ -1655,6 +1655,7 @@ static NTSTATUS wg_parser_connect(void *args)
     struct wg_parser *parser = params->parser;
     const WCHAR *uri = params->uri;
     bool use_mediaconv = false;
+    GstClock *clock;
     unsigned int i;
     int ret;
 
@@ -1688,7 +1689,8 @@ static NTSTATUS wg_parser_connect(void *args)
     gst_pad_set_event_function(parser->my_src, src_event_cb);
     gst_pad_set_element_private(parser->my_src, parser);
 
-    gst_element_set_clock(parser->container, gst_system_clock_obtain());
+    clock = gst_system_clock_obtain();
+    gst_element_set_clock(parser->container, clock);
 
     parser->start_offset = parser->next_offset = parser->stop_offset = 0;
     parser->next_pull_offset = 0;
@@ -1697,6 +1699,7 @@ static NTSTATUS wg_parser_connect(void *args)
     if (!parser->init_gst(parser))
         goto out;
 
+    gst_element_set_base_time(parser->container, gst_clock_get_time(clock));
     gst_element_set_state(parser->container, GST_STATE_PAUSED);
     ret = gst_element_get_state(parser->container, NULL, NULL, -1);
 
-- 
2.43.2

