From 75404f612e1da0f85540fe8ad4ebd22cf7d804c3 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Mon, 25 Mar 2024 09:58:04 +0100
Subject: [PATCH 16/39] === Marker commit, do not put into MR ===

---
 MAINTAINERS | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index 59873f6804c..20d715a2088 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -436,3 +436,5 @@ P:	Zebediah Figura <z.figura12@gmail.com>
 P:	Paul Gofman <gofmanp@gmail.com>
 P:	Erich E. Hoover <erich.e.hoover@wine-staging.com>
 W:	https://wine-staging.com/
+
+tmp
-- 
2.44.0

