From 9a1a82a504825d2656cbf13de31c5dd04534dec8 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sun, 10 Mar 2024 02:35:57 +0100
Subject: [PATCH 20/29] winegstreamer: Do not put pipeline back into PAUSED
 state.

---
 dlls/winegstreamer/wg_parser.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/winegstreamer/wg_parser.c b/dlls/winegstreamer/wg_parser.c
index 962cc136173..439977a4897 100644
--- a/dlls/winegstreamer/wg_parser.c
+++ b/dlls/winegstreamer/wg_parser.c
@@ -2110,13 +2110,13 @@ static NTSTATUS wg_parser_connect(void *args)
     }
     while (!parser_no_more_pads(parser) && !parser->error)
         pthread_cond_wait(&parser->init_cond, &parser->mutex);
-    if (parser->uri)
+    /*if (parser->uri)
     {
         gst_element_set_state(parser->container, GST_STATE_PAUSED);
         ret = gst_element_get_state(parser->container, NULL, NULL, -1);
         if (ret == GST_STATE_CHANGE_FAILURE)
             goto out;
-    }
+    }*/
     if (parser->error)
     {
         if (!parser->use_mediaconv && parser->init_gst == decodebin_parser_init_gst)
-- 
2.44.0

