From 6e45713173b6aa967c38105235bda6d43ddd4870 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Tue, 19 Mar 2024 05:07:11 +0100
Subject: [PATCH 23/29] winegstreamer: Discard buffers with negative PTS.

---
 dlls/winegstreamer/wg_parser.c | 57 +++++++++++++++++++---------------
 1 file changed, 32 insertions(+), 25 deletions(-)

diff --git a/dlls/winegstreamer/wg_parser.c b/dlls/winegstreamer/wg_parser.c
index 5c75776f523..9996f6848cd 100644
--- a/dlls/winegstreamer/wg_parser.c
+++ b/dlls/winegstreamer/wg_parser.c
@@ -307,6 +307,31 @@ static GstBuffer *wait_parser_stream_buffer(struct wg_parser *parser, struct wg_
     return buffer;
 }
 
+static void release_buffer(struct wg_parser *parser, struct wg_parser_stream *stream)
+{
+    if (stream->buffer)
+    {
+        gst_buffer_unmap(stream->buffer, &stream->map_info);
+        gst_buffer_unref(stream->buffer);
+        stream->buffer = NULL;
+    }
+
+    if (stream->buffer_queue && stream->buffer_queue_length)
+    {
+        GstBuffer *buffer;
+        do
+        {
+            buffer = *stream->buffer_queue;
+            stream->buffer_queue_length -= 1;
+            memmove(stream->buffer_queue, stream->buffer_queue + 1,
+                    stream->buffer_queue_length * sizeof(*stream->buffer_queue));
+        } while (!gst_buffer_map(buffer, &stream->map_info, GST_MAP_READ));
+        stream->buffer = buffer;
+    }
+
+    pthread_cond_broadcast(&parser->stream_event_cond);
+}
+
 static NTSTATUS wg_parser_stream_get_buffer(void *args)
 {
     const struct wg_parser_stream_get_buffer_params *params = args;
@@ -318,6 +343,7 @@ static NTSTATUS wg_parser_stream_get_buffer(void *args)
 
     pthread_mutex_lock(&parser->mutex);
 
+retry:
     if (stream)
         buffer = wait_parser_stream_buffer(parser, stream);
     else
@@ -374,9 +400,11 @@ static NTSTATUS wg_parser_stream_get_buffer(void *args)
                 return S_FALSE;
             }
             if (ret < 0)
-                wg_buffer->pts = -(gint64)stream_time / 100;
-            else
-                wg_buffer->pts = stream_time / 100;
+            {
+                release_buffer(parser, stream);
+                goto retry;
+            }
+            wg_buffer->pts = stream_time / 100;
         }
     }
     if ((wg_buffer->has_duration = GST_BUFFER_DURATION_IS_VALID(buffer)))
@@ -421,29 +449,8 @@ static NTSTATUS wg_parser_stream_release_buffer(void *args)
     struct wg_parser *parser = stream->parser;
 
     pthread_mutex_lock(&parser->mutex);
-
-    if (stream->buffer)
-    {
-        gst_buffer_unmap(stream->buffer, &stream->map_info);
-        gst_buffer_unref(stream->buffer);
-        stream->buffer = NULL;
-    }
-
-    if (stream->buffer_queue && stream->buffer_queue_length)
-    {
-        GstBuffer *buffer;
-        do
-        {
-            buffer = *stream->buffer_queue;
-            stream->buffer_queue_length -= 1;
-            memmove(stream->buffer_queue, stream->buffer_queue + 1,
-                    stream->buffer_queue_length * sizeof(*stream->buffer_queue));
-        } while (!gst_buffer_map(buffer, &stream->map_info, GST_MAP_READ));
-        stream->buffer = buffer;
-    }
-
+    release_buffer(parser, stream);
     pthread_mutex_unlock(&parser->mutex);
-    pthread_cond_broadcast(&parser->stream_event_cond);
 
     return S_OK;
 }
-- 
2.44.0

