From 5d78cdd2ad2e28d6bba872c7d01a3fd7c35b0922 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Mon, 25 Mar 2024 09:58:04 +0100
Subject: [PATCH 36/60] === Marker commit, do not put into MR ===

---
 MAINTAINERS | 1 +
 1 file changed, 1 insertion(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index ef3b4eabadb..63e3c5c147b 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -419,3 +419,4 @@ P:	Erich E. Hoover <erich.e.hoover@wine-staging.com>
 W:	https://wine-staging.com/
 
 seek
+tmp
-- 
2.53.0

