From 1df84dc25f87356918cf5a500e2b3119913de867 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Mon, 25 Mar 2024 09:58:04 +0100
Subject: [PATCH 36/49] === Marker commit, do not put into MR ===

---
 MAINTAINERS | 1 +
 1 file changed, 1 insertion(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index 2a76a529c7f..66055832c4d 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -418,3 +418,4 @@ P:	Erich E. Hoover <erich.e.hoover@wine-staging.com>
 W:	https://wine-staging.com/
 
 seek
+tmp
-- 
2.51.0

