From aea9312c603b0e820d910274ab3a66bdd41bf17a Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Thu, 18 Dec 2025 22:44:29 +0100
Subject: [PATCH 49/62] ntdll: Factor out bit search from RtlFindClearBits.

---
 dlls/ntdll/rtlbitmap.c | 49 ++++++++++++++++++++++++------------------
 1 file changed, 28 insertions(+), 21 deletions(-)

diff --git a/dlls/ntdll/rtlbitmap.c b/dlls/ntdll/rtlbitmap.c
index 4bf028fbe36..801c2720adc 100644
--- a/dlls/ntdll/rtlbitmap.c
+++ b/dlls/ntdll/rtlbitmap.c
@@ -256,6 +256,28 @@ ULONG WINAPI RtlFindSetBits(PCRTL_BITMAP lpBits, ULONG ulCount, ULONG ulHint)
   return ~0U;
 }
 
+static inline ULONG find_clear_bits( const RTL_BITMAP *bitmap, ULONG end, ULONG count, ULONG start )
+{
+  ULONG end = bitmap->SizeOfBitMap;
+
+  while (start < end)
+  {
+    /* FIXME: This could be made a _lot_ more efficient */
+    if (RtlAreBitsClear(bitmap, start, count))
+      return start;
+
+    /* Start from the beginning if we hit the end and started from ulHint */
+    if (start == end - 1 && ulHint)
+    {
+      end = ulHint;
+      start = ulHint = 0;
+    }
+    else
+      start++;
+  }
+  return ~0U;
+}
+
 /*************************************************************************
  * RtlFindClearBits	[NTDLL.@]
  *
@@ -271,36 +293,21 @@ ULONG WINAPI RtlFindSetBits(PCRTL_BITMAP lpBits, ULONG ulCount, ULONG ulHint)
  */
 ULONG WINAPI RtlFindClearBits(PCRTL_BITMAP lpBits, ULONG ulCount, ULONG ulHint)
 {
-  ULONG ulPos, ulEnd;
+  ULONG ulRet;
 
   TRACE("(%p,%lu,%lu)\n", lpBits, ulCount, ulHint);
 
   if (!lpBits || !ulCount || ulCount > lpBits->SizeOfBitMap)
     return ~0U;
 
-  ulEnd = lpBits->SizeOfBitMap;
-
   if (ulHint + ulCount > lpBits->SizeOfBitMap)
     ulHint = 0;
 
-  ulPos = ulHint;
-
-  while (ulPos < ulEnd)
-  {
-    /* FIXME: This could be made a _lot_ more efficient */
-    if (RtlAreBitsClear(lpBits, ulPos, ulCount))
-      return ulPos;
-
-    /* Start from the beginning if we hit the end and started from ulHint */
-    if (ulPos == ulEnd - 1 && ulHint)
-    {
-      ulEnd = ulHint;
-      ulPos = ulHint = 0;
-    }
-    else
-      ulPos++;
-  }
-  return ~0U;
+  ulRet = find_clear_bits(lpBits, lpBits->SizeOfBitMap, ulCount, ulHint);
+  /* Start from the beginning if we hit the end */
+  if (ulRet == ~0U && ulRet == lpBits->SizeOfBitMap - 1 && ulHint)
+    ulRet = find_clear_bits(lpBits, ulHint, ulCount, 0);
+  return ulRet;
 }
 
 /*************************************************************************
-- 
2.53.0

