From 4358573192b9e77b89cdeedb85bb9a0a4d804660 Mon Sep 17 00:00:00 2001
From: Reyka Matthies <openglfreak@googlemail.com>
Date: Thu, 16 Apr 2026 07:34:37 +0200
Subject: [PATCH 2/2] winegstreamer: Fix hang in IMFMediaSource::Shutdown.

---
 dlls/winegstreamer/media_source.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/dlls/winegstreamer/media_source.c b/dlls/winegstreamer/media_source.c
index 2f73496c114..2287b209fdf 100644
--- a/dlls/winegstreamer/media_source.c
+++ b/dlls/winegstreamer/media_source.c
@@ -1722,6 +1722,8 @@ static HRESULT WINAPI media_source_Pause(IMFMediaSource *iface)
 static HRESULT WINAPI media_source_Shutdown(IMFMediaSource *iface)
 {
     struct media_source *source = impl_from_IMFMediaSource(iface);
+    size_t queue_count = 0, i;
+    DWORD *queues = NULL;
 
     TRACE("%p.\n", iface);
 
@@ -1756,13 +1758,15 @@ static HRESULT WINAPI media_source_Shutdown(IMFMediaSource *iface)
     if (source->byte_stream)
         IMFByteStream_Close(source->byte_stream);
 
+    queues = calloc(source->stream_count, sizeof(DWORD));
     while (source->stream_count--)
     {
         struct media_stream *stream = source->streams[source->stream_count];
         IMFStreamDescriptor_Release(source->descriptors[source->stream_count]);
         IMFMediaEventQueue_QueueEventParamVar(stream->event_queue, MEError, &GUID_NULL, MF_E_SHUTDOWN, NULL);
         IMFMediaEventQueue_Shutdown(stream->event_queue);
-        MFUnlockWorkQueue(stream->async_commands_queue);
+        queues[queue_count++] = stream->async_commands_queue;
+        stream->async_commands_queue = 0;
         IMFMediaStream_Release(&stream->IMFMediaStream_iface);
     }
     free(source->descriptors);
@@ -1770,6 +1774,9 @@ static HRESULT WINAPI media_source_Shutdown(IMFMediaSource *iface)
 
     LeaveCriticalSection(&source->cs);
 
+    for (i = 0; i < queue_count; i++)
+        MFUnlockWorkQueue(queues[i]);
+    free(queues);
     MFUnlockWorkQueue(source->async_commands_queue);
 
     return S_OK;
-- 
2.53.0

